### ISSUE TITLE
How MultiView10 image transform works

### ISSUE THREAD

**Issue by patykov** [Microsoft/CNTK#2528]
I'm having some trouble understanding how the MultiView10 works, and was wondering if someone could clarify this.
I though this crop type would return 10 times the requested minibatch size since it's supposed to create 10 diferent crops from every single image: (4 corners + center) + (4 corners + center)flipped.

As I was testing some evaluation options, I noticed that the MultiView10 doesn't return 10 images, but only one. So, how exactly does it works? What transformation is applied to the images?

**Comment by calledbymountains** [comment]
Could you kindly clarify it ? (possibly with a code)  
I have the same concept about multiview10. 
It would be helpful to know how it behaves in your script.

**Comment by patykov** [comment]
For instance, in this example I load 2 images and compute the square error between than. 
For each minibatch I receive something like: 
`{Input('Input3', [#], [3 x 224 x 224]): MinibatchData(data=Value([3 x 1 x 3 x 224 x 224], CPU), samples=3, seqs=3), Input('Input4', [#], [3 x 224 x 224]): MinibatchData(data=Value([3 x 1 x 3 x 224 x 224], CPU), samples=3, seqs=3)}` 

But I was expecting something like:
`{Input('Input3', [#], [3 x 224 x 224]): MinibatchData(data=Value([30 x 1 x 3 x 224 x 224], CPU), samples=3, seqs=3), Input('Input4', [#], [3 x 224 x 224]): MinibatchData(data=Value([30 x 1 x 3 x 224 x 224], CPU), samples=3, seqs=3)}`

I don't understand how the 10 crops of the same image are returned. 

`def create_reader(map_file1, map_file2):
    transforms = [xforms.crop(crop_type='multiview10', crop_size=224),
                  xforms.scale(width=224, height=224, channels=3, interpolations='linear')]
    source1 = C.io.ImageDeserializer(map_file1, C.io.StreamDefs(
        source_image = C.io.StreamDef(field='image', transforms=transforms)))
    source2 = C.io.ImageDeserializer(map_file2, C.io.StreamDefs(
        target_image = C.io.StreamDef(field='image', transforms=transforms)))
    return C.io.MinibatchSource([source1, source2], max_sweeps=1, randomize=False)

x = C.input_variable((3,224,224))
y = C.input_variable((3,224,224))
z = C.squared_error(x, y)

reader = create_reader(mapFile1, mapFile2)

minibatch_size = 3
epoch_size = 9

input_map={
    x: reader.streams.source_image,
    y: reader.streams.target_image
}

sample_count = 0
while(sample_count < epoch_size):
    data=reader.next_minibatch(minibatch_size, input_map=input_map)
    print(data)
    sample_count += data[x].num_samples
    results = z.eval(data)
    `

**Comment by eldakms** [comment]
As correctly pointed out by calledbymountains, crop does not change the number of samples per minibatch, it only changes what are the images present in each sweep. So, if you have N images in your corpus:
X1, ... Xn
and apply the multiview, the corpus becomes (10x):
X11, X110, X2 .... XN10.
After that the images are randomized.(In case of randomization = false, all X11, X110 will be sequential (in sequential minibatches), but in the randomized case the whole corpus will be randomized, so this won't be the case.)
Then the minibatch is created based on the next minibatch_size images. The size of the minibatch has nothing to do with the cropping type.
Thanks!

**Comment by patykov** [comment]
So, if I want to evaluate all images I should use `max_sweeps=1` and update my condition in the evaluation loop to:
```
sample_count = 0
while(sample_count < epoch_size*10):
    data=reader.next_minibatch(minibatch_size, input_map=input_map)
    print(data)
    sample_count += data[x].num_samples
    results = z.eval(data)
```
?

Because I tried it that but after `epoch_size` samples I start receiving empty minibatches from `next_minibatch()`. 
What is the correct way to evaluate all `epoch_size*10` samples?

**Comment by calledbymountains** [comment]
I have some thoughts but it would be easier if you could test the following first and let us know if it changes something for you : 

- Don't specify `max_sweeps` option when constructing the reader. It should not hurt since you anyways know the sample size of your data by counting the number of lines in the `map_file`.

Now test your code using the `while(sample_count < epoch_size*10):` snippet and tell if it works now.

**Comment by cha-zhang** [comment]
@patykov have you tried to use max_sweeps = 10?

**Comment by patykov** [comment]
Using a trained network I've done these tests:

- I didn't specified `max_sweeps` and used `while(sample_count < epoch_size*10):` in the evaluation loop. What I got was the exactly same results repeated 10 times.

- Specified `max_sweeps=10` and used `while(sample_count < epoch_size*10):` in the evaluation loop.  And again, I got the exactly same results repeated 10 times.

If the crops were working as expected it should return at least slightly different results, right? 
Is there another way to test it?
Has anybody been able to evaluate all `epoch_size*10` samples correctly?

**Comment by calledbymountains** [comment]
Do the following for a definitive check : 
Read data from the minibatch and write it to an image file. This way see if the images read correspond to different crops. That should be a definitive testing. If the trained network is giving the same accuracy on all the crops that does not tell something definitive about the cropping, in my opinion

**Comment by patykov** [comment]
I've repeated the same tests as before, but saving each image to check the crop as suggested by @calledbymountains . 
All the crops are the same! Apparently it is only doing the top left crop.

**Comment by calledbymountains** [comment]
@patykov 
I am going to test it now. It would be faster if you could post your snippet for testing the crops as I suggested so that I could test it quicker. Thanks

**Comment by patykov** [comment]
Sure:

```
def create_reader(map_file, num_classes):
    transforms = [xforms.crop(crop_type='multiview10', crop_size=224),
                  xforms.scale(width=224, height=224, channels=3, interpolations='linear')]

    return C.io.MinibatchSource(C.io.ImageDeserializer(map_file,C.io. StreamDefs(
                features = C.io.StreamDef(field='image', transforms=transforms),
                labels   = C.io.StreamDef(field='label', shape=num_classes))), 
                randomize=False, max_sweeps=10)


reader = create_reader(mapFile, num_classes)

features_si = reader['features']
sample_count = 0
while(sample_count < epoch_size*10):
    data=reader.next_minibatch(mb_size)
    img_array = np.rollaxis(np.squeeze(data[features_si].as_sequences()), 0, 3)
    cv2.imwrite(os.path.join(output_dir, '{}.jpeg'.format(sample_count)), img_array)
    sample_count += data[features_si].num_samples
```

**Comment by robertomest** [comment]
Are there any updates on this issue? I've also tested and the multiview10 crop does not seem to be working as intended.

**Comment by patykov** [comment]
@calledbymountains did you have the chance to test it? Any updates on this issue?

**Comment by calledbymountains** [comment]
@patykov  Yes. I did. It indeed does not work correctly. The detailed stats are on my lab computer. I am on a medical leave and will return back to the lab on Monday. I shall update the details right on Monday. I am truly sorry but I should have informed much before
